lib/incoming_mail/handler.rb

```diff
@@ -63,19 +63,25 @@ def create_comment(msg)
       proposal = find_proposal(find_public_id(msg)) or return
       comment_text = find_comment_text(msg)
       comment_user = find_comment_user(msg)
-      comment = nil
-      if proposal.has_subscriber?(comment_user)
-        # already in the loop, just add comment.
-        comment = proposal.comments.create(comment_text: comment_text, user: comment_user)
-      else
-        # yes, user adds self as observer, which also generates comment
-        proposal.add_observer(comment_user, comment_user, comment_text)
-        comment = proposal.comments.last
+      # IMPORTANT that we check/add as observer before we create comment,
+      # since comment will create as a user if not already,
+      # and we want the reason logged.
+      unless proposal.has_subscriber?(comment_user)
+        add_user_as_observer(comment_user, proposal)
       end
+      comment = proposal.comments.create(comment_text: comment_text, user: comment_user)
       Dispatcher.on_comment_created(comment) # sends email
       comment
     end
 
+    def add_user_as_observer(user, proposal)
+      observer_role = Role.find_or_create_by(name: 'observer')
+      observation = Observation.new(user_id: user.id, role_id: observer_role.id, proposal_id: proposal.id)
+      proposal.observations << observation
+      Dispatcher.on_observer_added(observation, "Added comment via email reply")
+      observation
+    end
```

Comment:
I was wrong. During my refactor, I went from creating 1 comment to creating 2 comments: 1 for the email reply, and one `update_comment` comment for the addition of the observer, instead of a single combined comment.

So yes, `Proposal.add_observer` is appropriate to re-use here.